packages/core/contracts/optimistic-assertor/implementation/examples/DataAsserter.sol

```diff
@@ -0,0 +1,113 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "../../interfaces/OptimisticAssertorInterface.sol";
+import "../../../common/implementation/AncillaryData.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+// This contract allows assertions on any form of data to be made using the UMA Optimistic Assertor and stores the
+// proposed value so that it may be retrieved on chain. It only enables one assertion per pair of dataId and asserter
+// address. The dataId is intended to be an arbitrary value that uniquely identifies a specific piece of information in
+// the consuming contract and is replaceable. Similarly, any data structure can be used to replace the asserted data.
+contract DataAsserter {
+    using SafeERC20 for IERC20;
+    IERC20 public immutable defaultCurrency;
+    OptimisticAssertorInterface public immutable oa;
+    uint256 public constant assertionLiveness = 7200;
+    bytes32 public immutable defaultIdentifier;
+
+    struct DataAssertion {
+        bytes32 dataId; // This could be an arbitrary data type.
+        uint256 data; // This could be an arbitrary data type.
+        address asserter; // The address that made the assertion.
+        bytes32 oaAssertionId; // The optimistic assertor assertion ID.
+    }
+
+    mapping(bytes32 => bytes32) public oaIdsToInternalIds;
+
+    mapping(bytes32 => bool) public assertionsResolved;
+
+    mapping(bytes32 => DataAssertion) public assertionsData;
+
+    event DataAsserted(bytes32 indexed dataId, uint256 data, address indexed asserter, bytes32 oaAssertionId);
+
+    event DataAssertionResolved(bytes32 indexed dataId, uint256 data, address indexed asserter, bytes32 oaAssertionId);
+
+    constructor(address _defaultCurrency, address _optimisticAssertor) {
+        defaultCurrency = IERC20(_defaultCurrency);
+        oa = OptimisticAssertorInterface(_optimisticAssertor);
+        defaultIdentifier = oa.defaultIdentifier();
+    }
+
+    // For a given dataId and asserter, returns a boolean indicating whether the data is accessible and the data itself.
+    function getData(bytes32 dataId, address asserter) public view returns (bool, uint256) {
+        bytes32 dataAssertionId = getAssertionId(dataId, asserter);
+        if (!assertionsResolved[dataAssertionId]) return (false, 0);
+        return (true, assertionsData[dataAssertionId].data);
+    }
+
+    // Asserts data for a specific dataId on behalf of an asserter address.
+    function assertDataFor(
+        bytes32 dataId,
+        uint256 data,
+        address asserter
```

Comment:
It might make sense to use the same pattern as in Optimistic Asserter where zero address is overriden with `msg.sender`

Reply:
Good point